Add style invalidity state for animations

anttijk opened the pull request:
#### c5a71dadf89f5816fedaab3d7efaaec4e247fdc6
<pre>
Add style invalidity state for animations
<a href="https://bugs.webkit.org/show_bug.cgi?id=264932">https://bugs.webkit.org/show_bug.cgi?id=264932</a>
<a href="https://rdar.apple.com/118500247">rdar://118500247</a>

Reviewed by Antoine Quint.

We may be able to apply animation effects only without reapplying normal style rules.

* Source/WebCore/animation/ElementAnimationRareData.h:
(WebCore::ElementAnimationRareData::hasPropertiesOverridenAfterAnimation const):
(WebCore::ElementAnimationRareData::setHasPropertiesOverridenAfterAnimation):

If we have rules the override animation effects we can not use this optimization.

* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::invalidateElement):

Set the animation invalidity state instead of invaliding the element fully.

(WebCore::KeyframeEffect::animationDidChangeTimingProperties):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::invalidateStyleForAnimation):
(WebCore::Element::hasPropertiesOverridenAfterAnimation const):
(WebCore::Element::setHasPropertiesOverridenAfterAnimation):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/Node.h:
* Source/WebCore/style/StyleInvalidator.cpp:
(WebCore::Style::Invalidator::invalidateIfNeeded):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::styleForStyleable):

Use the base style from the previous frame instead of computing it again if only
the animation is invalid.

(WebCore::Style::TreeResolver::createAnimatedElementUpdate):
(WebCore::Style::TreeResolver::determineResolutionType):
(WebCore::Style::shouldResolvePseudoElement): Deleted.
* Source/WebCore/style/StyleTreeResolver.h:
* Source/WebCore/style/StyleValidity.h:
* Source/WebCore/style/Styleable.h:
(WebCore::Styleable::hasPropertiesOverridenAfterAnimation const):
(WebCore::Styleable::setHasPropertiesOverridenAfterAnimation const):
* Source/WebCore/testing/Internals.cpp:
(WebCore::styleValidityToToString):

Canonical link: <a href="https://commits.webkit.org/270890@main">https://commits.webkit.org/270890@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f31bbd48cdea8dfaf1e0b60a40cf0471ccd3dff3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26734 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5348 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27974 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28946 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24444 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/27179 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7179 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2745 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/24353 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26996 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4168 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22953 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3662 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/3705 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23945 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29431 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24381 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24345 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/29967 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3741 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1936 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27868 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5190 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6421 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4210 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4093 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on Source/WebCore/animation/KeyframeEffect.cpp:
```diff
@@ -1849,7 +1849,8 @@ void KeyframeEffect::animationDidChangeTimingProperties()
     updateAcceleratedAnimationIfNecessary();
-    invalidate();
+    if (auto target = targetStyleable())
+        target->element.invalidateStyleInternal();
```
Why does that not use `invalidateStyleForAnimation()` like `invalidate()` does?

anttijk replied:
No reason!

anttijk replied:
Turns out it was fixing something but the right place for the full invalidation is in KeyframeEffect::setKeyframes. New keyframes may change override interaction with the base style.

anttijk committed:
Add style invalidity state for animations
https://bugs.webkit.org/show_bug.cgi?id=264932
rdar://118500247

Reviewed by Antoine Quint.

We may be able to apply animation effects only without reapplying normal style rules.

* Source/WebCore/animation/ElementAnimationRareData.h:
(WebCore::ElementAnimationRareData::hasPropertiesOverridenAfterAnimation const):
(WebCore::ElementAnimationRareData::setHasPropertiesOverridenAfterAnimation):

If we have rules the override animation effects we can not use this optimization.

* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::invalidateElement):

Set the animation invalidity state instead of invaliding the element fully.

(WebCore::KeyframeEffect::animationDidChangeTimingProperties):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::invalidateStyleForAnimation):
(WebCore::Element::hasPropertiesOverridenAfterAnimation const):
(WebCore::Element::setHasPropertiesOverridenAfterAnimation):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/Node.h:
* Source/WebCore/style/StyleInvalidator.cpp:
(WebCore::Style::Invalidator::invalidateIfNeeded):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::styleForStyleable):

Use the base style from the previous frame instead of computing it again if only
the animation is invalid.

(WebCore::Style::TreeResolver::createAnimatedElementUpdate):
(WebCore::Style::TreeResolver::determineResolutionType):
(WebCore::Style::shouldResolvePseudoElement): Deleted.
* Source/WebCore/style/StyleTreeResolver.h:
* Source/WebCore/style/StyleValidity.h:
* Source/WebCore/style/Styleable.h:
(WebCore::Styleable::hasPropertiesOverridenAfterAnimation const):
(WebCore::Styleable::setHasPropertiesOverridenAfterAnimation const):
* Source/WebCore/testing/Internals.cpp:
(WebCore::styleValidityToToString):

Canonical link: https://commits.webkit.org/270890@main
